streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java

```diff
@@ -806,7 +966,15 @@ public void process(final ProcessorSupplier<? super K, ? super V> processorSuppl
 
         final Set<String> allSourceNodes = ensureJoinableWith((AbstractStream<K, VO>) other);
 
-        final String name = builder.newProcessorName(leftJoin ? LEFTJOIN_NAME : JOIN_NAME);
+        final JoinedInternal<K, V, VO> joinedInternal = new JoinedInternal<>(joined);
+        final NamedInternal renamed = new NamedInternal(joinedInternal.name());
+        // we actually do not need to generate processor names for join operation if specified name.
+        // But, we still need to burn index for join operation keep topology backward compatibility.
+        if (renamed.isDefined()) {
```

Comment:
@fhussonnois can we move the logic from here to the `NamedInternal` to always generate the name to ensure we increment the counter for backward compatibility?  

Something along the lines of 
```java
private String orElseGet(final Supplier<String> supplier) {
        final String generatedName = supplier.get();
        return Optional.ofNullable(this.name).orElseGet(() -> generatedName);
    }
```